### PR TITLE
Feature/activity on board since last online 93646616

### PR DESCRIPTION
**Pull request by sueyeh**
Pivotal story: https://www.pivotaltracker.com/story/show/93646616

changes-since-last-online

We save the lastOnlineAt on collaborators every time a user joins a board, leaves a board, or heartbeats. When we compare lastOnlineAt with the last time a card is updated, we add a 10 second buffer because the dashboard loads faster than it takes to update the lastOnlineAt (such as when a user is leaving a board)

Comments and thoughts appreciated!

### CODE REVIEW

File: lib/models/board.coffee [hunk 1]
```diff
@@ -26,8 +27,8 @@ BoardSchema.virtual('currentUserId').get () -> @vCurrentUserId
 BoardSchema.virtual('currentUserId').set (currentUserId) -> @vCurrentUserId = currentUserId
 
 BoardSchema.pre 'save', (next) ->
+  @updated = new Date() unless @modifiedPaths().length == 0 or @isModified('collaborators')
   @created = new Date() unless @created?
-  @updated = new Date()
   @collaborate @creator
   next()
```
Comment (mwynholds) [inline lib/models/board.coffee]: I just thought of something...  On line 30 we are *not* setting `@updated` if the collaborators property is modified.  But what if the collaborators property is modified *and* the board name is modified (for example)?  We should actually be checking to see if the collaborators property is the only one that is modified.

And we should probably pull that logic in to its own function.

File: lib/services/sockets/user.coffee [hunk 2]
```diff
@@ -29,6 +29,8 @@ class UserHandler
           return ( logger.error -> "Error finding user #{@userId} and board #{@boardId}: #{err}" ) if err?
           return ( logger.error -> "Cannot find user: #{@userId}" ) unless user?
           return ( logger.error -> "Cannot find board: #{@boardId} (user: #{user.activeIdentity.profile()})" ) unless board?
+          board.updateLastOnline(user)
+          board.save()
```
Comment (mwynholds) [inline lib/services/sockets/user.coffee]: I feel like we should have a callback here that at least checks to see if an error happens and logs it.

File: lib/services/sockets/user.coffee [hunk 3]
```diff
@@ -48,6 +50,8 @@ class UserHandler
...
```
Comment (mwynholds) [inline lib/services/sockets/user.coffee]: Same as line 33 above.

File: lib/services/sockets/user.coffee [hunk 4]
```diff
@@ -59,6 +63,8 @@ class UserHandler
...
```
Comment (mwynholds) [inline lib/services/sockets/user.coffee]: Same as line 33 above.

File: lib/models/board.coffee [hunk 5]
```diff
@@ -26,8 +27,8 @@ BoardSchema.virtual('currentUserId').get () -> @vCurrentUserId
 BoardSchema.virtual('currentUserId').set (currentUserId) -> @vCurrentUserId = currentUserId
 
 BoardSchema.pre 'save', (next) ->
+  @updated = new Date() unless @modifiedPaths().length == 0 or @isModified('collaborators')
```
Comment (mwynholds) [inline lib/models/board.coffee]: I just thought of something...  On line 30 we are *not* setting `@updated` if the collaborators property is modified.  But what if the collaborators property is modified *and* the board name is modified (for example)?  We should actually be checking to see if the collaborators property is the only one that is modified.

And we should probably pull that logic in to its own function.

File: assets/css/modules/board-cell.less [hunk 6]
```diff
@@ -124,7 +124,10 @@ Styleguide: Board.Cell
 
 .board-cell__updated-at {
   text-align: right;
-  color: @redActive;
+
+  &.board-cell__updated-at--unseen {
```
Comment (rudyjahchan) [inline assets/css/modules/board-cell.less]: Is this standard naming conventions for CSS classes?

File: lib/controllers/boards.coffee [hunk 7]
```diff
@@ -27,7 +27,6 @@ class BoardsController extends ApplicationController
         boardModel.save (err, _) ->
           ( logger.error -> "Error saving board '#{boardModel.name}' - #{err}" ) if err?
 
-
```
Comment (rudyjahchan) [inline lib/controllers/boards.coffee]: Are you sure this empty line wasn't important? This is CoffeeScript after all (that one was for @sidke)

File: assets/css/modules/board-cell.less (hunk 6 above)
Comment (nfiniteset) [inline assets/css/modules/board-cell.less]: Looks like BEM syntax to me. :+1: 

Is the `&` necessary though? I would expect this selector to work without it as long as it's declared after `.board-cell__updated-at`.

File: lib/services/sockets/user.coffee (hunk 2 above)
Comment (rudyjahchan) [inline lib/services/sockets/user.coffee]: Perfect time to try out promises ;-)

File: assets/css/modules/board-cell.less (hunk 6 above)
Comment (sueyeh) [inline assets/css/modules/board-cell.less]: noted!

File: lib/controllers/boards.coffee (hunk 7 above)
Comment (sueyeh) [inline lib/controllers/boards.coffee]: I'll make sure to add the blank line back in

File: lib/services/sockets/user.coffee [hunk 8]
```diff
@@ -29,6 +29,9 @@ class UserHandler
           return ( logger.error -> "Error finding user #{@userId} and board #{@boardId}: #{err}" ) if err?
           return ( logger.error -> "Cannot find user: #{@userId}" ) unless user?
           return ( logger.error -> "Cannot find board: #{@boardId} (user: #{user.activeIdentity.profile()})" ) unless board?
+          board.updateLastOnline(user)
+          board.save().then null, (err) ->
```
Comment (rudyjahchan) [inline lib/services/sockets/user.coffee]: `catch` instead of `then`

```
board.save().catch (err) ->
```

And `save` returns a promise?

Comment (sueyeh) [inline lib/services/sockets/user.coffee]: It does, but there is no "catch" https://github.com/Automattic/mongoose/blob/4.0.6/lib/promise.js

**Pull request merged**